src/Common/HostResolvePool.cpp

```diff
@@ -223,7 +227,10 @@ void HostResolver::updateImpl(Poco::Timestamp now, std::vector<Poco::Net::IPAddr
         {
             CurrentMetrics::add(metrics.active_count, 1);
             ProfileEvents::increment(metrics.discovered, 1);
-            merged.push_back(Record(*it_next, now));
+            if (merged.empty() || merged.back().address != *it_next)
+                merged.push_back(Record(*it_next, now));
+            else
+                merged.back().resolve_time = now;
```

Comment:
I do not understand that lines.
It is a case when new address discovered. Why do you do additional check here `if (merged.empty() || merged.back().address != *it_next)`? This condition is always true here.

Reply:
On my host getaddrinfo returns 3 record per each address, for example
```
(gdb) frame
#0  Poco::Net::DNS::hostByName (hostname=..., hintFlags=<optimized out>) at ./build-clang-17/./base/poco/Net/src/DNS.cpp:80
80		if (rc == 0)
(gdb) l
75		struct addrinfo* pAI;
76		struct addrinfo hints;
77		std::memset(&hints, 0, sizeof(hints));
78		hints.ai_flags = hintFlags;
79		int rc = getaddrinfo(hostname.c_str(), NULL, &hints, &pAI);
80		if (rc == 0)
81		{
82			HostEntry result(pAI);
83			freeaddrinfo(pAI);
84			return result;
(gdb) p *pAI
$14 = {ai_flags = 32, ai_family = 2, ai_socktype = 1, ai_protocol = 6, ai_addrlen = 16, ai_addr = 0x76bbf2316170, ai_canonname = 0x0, ai_next = 0x76bbf2316180}
(gdb) p *pAI->ai_next
$15 = {ai_flags = 32, ai_family = 2, ai_socktype = 2, ai_protocol = 17, ai_addrlen = 16, ai_addr = 0x76bbf23161b0, ai_canonname = 0x0, ai_next = 0x76bbf23161c0}
(gdb) p *pAI->ai_next->ai_next
$16 = {ai_flags = 32, ai_family = 2, ai_socktype = 3, ai_protocol = 0, ai_addrlen = 16, ai_addr = 0x76bbf23161f0, ai_canonname = 0x0, ai_next = 0x0}
(gdb) p *(sockaddr_in*)pAI->ai_addr
$19 = {sin_family = 2, sin_port = 0, sin_addr = {s_addr = 16777343}, sin_zero = "\000\000\000\000\000\000\000"}
(gdb) p *(sockaddr_in*)pAI->ai_next->ai_addr
$20 = {sin_family = 2, sin_port = 0, sin_addr = {s_addr = 16777343}, sin_zero = "\000\000\000\000\000\000\000"}
(gdb) p *(sockaddr_in*)pAI->ai_next->ai_next->ai_addr
$21 = {sin_family = 2, sin_port = 0, sin_addr = {s_addr = 16777343}, sin_zero = "\000\000\000\000\000\000\000"}
```
(16777343 ==> '127.0.0.1')
so each IP has three copies in `next_gen` parameter in `HostResolver::updateImpl` method.

Reply:
By the way, randomizing in HostResolver does not respect address priority from /etc/gai.conf

Reply:
I did not saw any code in Clickhouse which cares about ip order, so I did not supported it here either. if you wish you could try to add it!

Reply:
> so each IP has three copies in next_gen parameter in HostResolver::updateImpl method.

I thought that DNSResolver eliminating duplicates. But if it does not, than some one should do it. Lets do it in HostResolver because it expects unique records right now. Thanks for that info.

Reply:
BTW you reminded me that it shoud be like that here https://github.com/ClickHouse/ClickHouse/blob/f75447bf11e76e09a291b3a8793f3e953bf9e279/src/Common/HostResolvePool.cpp#L50

```
- , resolve_function([](const String & host_to_resolve) { return DNSResolver::instance().resolveHostAll(host_to_resolve); })
+ , resolve_function([](const String & host_to_resolve) { return DNSResolver::instance().resolveHostAllInOriginOrder(host_to_resolve); })  
```

Reply:
Second random in choosing IPs by weights, so ordering here does not fix it.